### ISSUE TITLE
Allow WHERE conditions to compare columns in same table

### ISSUE THREAD

**Issue by five07** [prisma/prisma#5048]
## Problem
Prisma should allow `WHERE` conditions to compare columns in same table

## Suggested solution
Expand `WHERE` filters to include columns of the same type from the same table

Simple example model
```
model Shipment {
  id               String @id @default(cuid())
  quantityShipped  Int
  quantityReceived Int
}
```

Raw SQL:
```
SELECT * FROM Shipment WHERE quantityReceived < quantityShipped;
```

Currently for example model:
```
type IntFilter = {
  equals?: number | null
  not?: number | IntFilter | null
  in?: Enumerable<number> | null
  notIn?: Enumerable<number> | null
  lt?: number | null
  lte?: number | null
  gt?: number | null
  gte?: number | null
}
```

Could alternately be:
```
type IntFilter = {
  equals?: number | ShipmentIntColumnInput | null
  not?: number | IntFilter | ShipmentIntColumnInput | null
  in?: Enumerable<number> | Enumerable<ShipmentIntColumnInput> | null
  notIn?: Enumerable<number> | Enumerable<ShipmentIntColumnInput> | null
  lt?: number | ShipmentIntColumnInput | null
  lte?: number | ShipmentIntColumnInput | null
  gt?: number | ShipmentIntColumnInput | null
  gte?: number | ShipmentIntColumnInput | null
}
```

## Alternatives
Currently only raw queries work

Thanks!

**Comment by vktrl** [comment]
Would like to have it as well. Right now I'm using DB functions that calculate and set extra columns so I can filter my queries.

**Comment by sctnightcore** [comment]
any updates?

**Comment by deviprsd** [comment]
I would like to point out that this feature should be able to handle use cases such as these (**note**: *the left join query in particular*)

```sql
SELECT 
    COUNT(g.id) AS 'total_used', 
    s.`product_name`, 
    s.`current_period_start` AS 'current_period_start', 
    s.`current_period_end` AS 'current_period_end'
FROM `subscription` s
    JOIN `user` u ON s.`customer_id` = u.`customer_id`
    LEFT JOIN `goat` g ON g.`user_id` = u.`id` AND
          g.`date_created` BETWEEN s.`current_period_start` AND s.`current_period_end`
    WHERE u.`id` = "00000-00000" AND s.`status` = 'ACTIVE'
```

**Comment by danshapir** [comment]
Any updates?
This is a super important feature..

**Comment by TomerAberbach** [comment]
Here's another use case for this: https://github.com/prisma/prisma/discussions/11646

**Comment by debrice** [comment]
I'm curious is this is gonna be ever be addressed ? I'm likely over-simplifying the issue but this seems like such a fundamental feature of relational database (amount payed < amount due, delivery received < delivery ordered, delivery date < expected delivery date...) that I can't figure out why it's not handled. I understand that it can be dealt with using raw query but then what is the value of Prisma if one has to resort to raw query for basic query.

**Comment by Suleyman1406** [comment]
Any update?

**Comment by volkankorkmaz** [comment]
Any updates? Do you have any ETA about it?

**Comment by DuriDuri** [comment]
Any updates? Do you have any ETA about it?

**Comment by Riz-waan** [comment]
I also would like upvote this request. This is an important feature to me.

**Comment by Ontegor** [comment]
I see that the developer @cprieto  is working on this task. Let's hope he solves this problem 👍

**Comment by matthewmueller** [comment]
Hey @Ontegor, yes, we started designing this feature. We'll post an update here once we have something worth sharing!

**Comment by volkankorkmaz** [comment]
> @Ontegor yep, we started designing this feature. We'll post an update here once we have something worth sharing!

Sounds good!

**Comment by shtse8** [comment]
any updates?

**Comment by volkankorkmaz** [comment]
Hi,

Do you have any ETA about it? 

Thx

**Comment by matthewmueller** [comment]
Hey folks 👋  

We started designing this feature and we’re ready to share our proposal. Please let us know what you think! I'd also like to invite you to [jump on a quick call with us](https://calendly.com/mueller-at-prisma/prisma-compare) to share your thoughts.

## Design

We aim to support type-safe field references on all filters (e.g. `where`, `having`, etc.):

Let’s consider the following `schema.prisma`:

```groovy
model Order {
  id     String @id
  paid   Int
  due    Int
  user   User   @relation(fields: [userId], references: [id])
  userId String
}

model User {
  id     String  @id
  email  String  @unique
  name   String
  tags   String[]
  orders Order[]
}
```

You can now express a query to find all the orders where `paid < due`.

```jsx
await prisma.order.findMany({
  where: {
    paid: {
      lt: prisma.order.fields.due // paid < due
    }
  }
})
```

This proposal adds a special `fields` property to every model that contains the list of fields for that model. Using the models above, fields contains the following properties:

```tsx
const { id, paid, due, userId } = prisma.order.fields
const { id, email, name, tags } prisma.user.fields
```

### Compile-Time Type-Safety

The Typescript compiler ensures that comparisons are only possible on fields of the same type. For example, the following will trigger an error while type-checking:

```jsx
await prisma.order.findMany({
  where: {
    id: { equals: prisma.order.fields.due }
    // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    // Type error: id is a string, while amountDue is for integers
  }
})
```

Comparisons are also currently restricted to being on the same model:

```jsx
await prisma.order.findMany({
  where: {
    id: { equals: prisma.user.fields.name }
    // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    // Type error: name is a field on the User model
  }
})
```

### Referencing Scalar Lists

For data sources that support scalar lists (e.g. PostgreSQL), you’ll be also able to reference scalar lists with the `in` and `notIn` filters, so you can now express “find all results where one field is within a list of other fields. For example,

```jsx
await prisma.user.findMany({
  where: {
    // find all users where name is within a list of tags
    name: { in: prisma.user.fields.tags }  
  }
})
```

We hope you’ll find a better use case for this feature than our example… 😅 Let us know!

## Limitations

Referencing fields will *only* be available in non-unique filters for now. For example, the following would not be possible:

```jsx
await prisma.user.findUnique({
  where: {
    email: prisma.user.fields.name
  }
})
```

**Comment by shtse8** [comment]
> Hey folks 👋
> 
> We've started designing this feature and we’re ready to share our proposal. Please let us know what you think! I'd also like to invite you to [jump on a quick call with us](https://calendly.com/mueller-at-prisma/prisma-compare) to share your thoughts.
> 
> ## Design
> We aim to support type-safe field references on all filters (e.g. `where`, `having`, etc.):
> 
> Let’s consider the following snippet from a `schema.prisma`:
> 
> ```groovy
> model Order {
>   id     String @id
>   paid   Int
>   due    Int
>   user   User   @relation(fields: [userId], references: [id])
>   userId String
> }
> 
> model User {
>   id     String  @id
>   email  String  @unique
>   name   String
>   tags   String[]
>   orders Order[]
> }
> ```
> 
> You can now express a query to find all the orders where `paid < due`.
> 
> ```js
> await prisma.order.findMany({
>   where: {
>     paid: {
>       lt: prisma.order.fields.due // paid < due
>     }
>   }
> })
> ```
> 
> This proposal adds a special `fields` property to every model that contains the list of fields for that model. Using the models above, fields contains the following properties:
> 
> ```tsx
> const { id, paid, due, userId } = prisma.order.fields
> const { id, email, name, tags } prisma.user.fields
> ```
> 
> ### Compile-Time Type-Safety
> The Typescript compiler ensures that comparisons are only possible on fields of the same type. For example, the following will trigger an error while type-checking:
> 
> ```js
> await prisma.order.findMany({
>   where: {
>     id: { equals: prisma.order.fields.due }
>     // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     // Type error: id is a string, while amountDue is for integers
>   }
> })
> ```
> 
> Comparisons are also currently restricted to being on the same model:
> 
> ```js
> await prisma.order.findMany({
>   where: {
>     id: { equals: prisma.user.fields.name }
>     // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     // Type error: name is a field on the User model
>   }
> })
> ```
> 
> ### Referencing Scalar Lists
> For data sources that support scalar lists (e.g. PostgreSQL), you’ll be also able to reference scalar lists with the `in` and `notIn` filters, so you can now express “find all results where one field is within a list of other fields. For example,
> 
> ```js
> await prisma.user.findMany({
>   where: {
>     // find all users where name is within a list of tags
>     name: { in: prisma.user.fields.tags }  
>   }
> })
> ```
> 
> We hope you’ll find a better use case for this feature than our example… 😅 Let us know!
> 
> ## Limitations
> Referencing fields will _only_ be available in non-unique filters for now. For example, the following would not be possible:
> 
> ```js
> await prisma.user.findUnique({
>   where: {
>     email: prisma.user.fields.name
>   }
> })
> ```

Good to see there is a type safety check. But what if I want to compare timestamp (int) and datetime (datetime) column in database?

**Comment by danshapir** [comment]
> Hey folks 👋  
> 
> We've started designing this feature and we’re ready to share our proposal. Please let us know what you think! I'd also like to invite you to [jump on a quick call with us](https://calendly.com/mueller-at-prisma/prisma-compare) to share your thoughts.
> 
> ## Design
> 
> We aim to support type-safe field references on all filters (e.g. `where`, `having`, etc.):
> 
> Let’s consider the following snippet from a `schema.prisma`:
> 
> ```groovy
> model Order {
>   id     String @id
>   paid   Int
>   due    Int
>   user   User   @relation(fields: [userId], references: [id])
>   userId String
> }
> 
> model User {
>   id     String  @id
>   email  String  @unique
>   name   String
>   tags   String[]
>   orders Order[]
> }
> ```
> 
> You can now express a query to find all the orders where `paid < due`.
> 
> ```jsx
> await prisma.order.findMany({
>   where: {
>     paid: {
>       lt: prisma.order.fields.due // paid < due
>     }
>   }
> })
> ```
> 
> This proposal adds a special `fields` property to every model that contains the list of fields for that model. Using the models above, fields contains the following properties:
> 
> ```tsx
> const { id, paid, due, userId } = prisma.order.fields
> const { id, email, name, tags } prisma.user.fields
> ```
> 
> ### Compile-Time Type-Safety
> 
> The Typescript compiler ensures that comparisons are only possible on fields of the same type. For example, the following will trigger an error while type-checking:
> 
> ```jsx
> await prisma.order.findMany({
>   where: {
>     id: { equals: prisma.order.fields.due }
>     // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     // Type error: id is a string, while amountDue is for integers
>   }
> })
> ```
> 
> Comparisons are also currently restricted to being on the same model:
> 
> ```jsx
> await prisma.order.findMany({
>   where: {
>     id: { equals: prisma.user.fields.name }
>     // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     // Type error: name is a field on the User model
>   }
> })
> ```
> 
> ### Referencing Scalar Lists
> 
> For data sources that support scalar lists (e.g. PostgreSQL), you’ll be also able to reference scalar lists with the `in` and `notIn` filters, so you can now express “find all results where one field is within a list of other fields. For example,
> 
> ```jsx
> await prisma.user.findMany({
>   where: {
>     // find all users where name is within a list of tags
>     name: { in: prisma.user.fields.tags }  
>   }
> })
> ```
> 
> We hope you’ll find a better use case for this feature than our example… 😅 Let us know!
> 
> ## Limitations
> 
> Referencing fields will *only* be available in non-unique filters for now. For example, the following would not be possible:
> 
> ```jsx
> await prisma.user.findUnique({
>   where: {
>     email: prisma.user.fields.name
>   }
> })
> ```

That looks great, but I'm unsure as to how it works with JOINs. Those are the more realistic scenarios of needing such a feature.
Moreover I think there should be a solution for complex comparison, even if it will be a small part directly written in SQL. Otherwise it will require writing everything in SQL.

**Comment by matthewmueller** [comment]
> But what if I want to compare timestamp (int) and datetime (datetime) column in database?

Thanks for your comment @shtse8. Do you have a use case in mind where you'd like to do this? I'd also be curious what the SQL alternative would look like for this if you have something in mind.

>  Those are the more realistic scenarios of needing such a feature.

Thanks for your comment @danshapir, can you share an example where you'd need to join with column references?

**Comment by danshapir** [comment]
> > But what if I want to compare timestamp (int) and datetime (datetime) column in database?
> 
> Thanks for your comment @shtse8. Do you have a use case in mind where you'd like to do this? I'd also be curious what the SQL alternative would look like for this if you have something in mind.
> 
> > Those are the more realistic scenarios of needing such a feature.
> 
> Thanks for your comment @danshapir, can you share an example where you'd need to join with column references?

@matthewmueller Sure.
Assume we have 2 tables.
Accounts and Transactions
We have a field on the Accounts called LinkTime, a datetime.
We want to find all transactions where their createdAt is larger than the LinkTime of their parent account.

That's for the simple query, just comparison of fields from a join.
For the more complex scenario, but something we all do a thousand times using SQL, we create a dynamic property as part of the select (computed of multiple fields or any other logic, or maybe aggregated SUM/MIN) and we want to only fetch items that are larger or equal to it.

**Comment by danshapir** [comment]
@SevInf How is this closed? The solution doesn't support JOIN queries... The basically main reason for this requirement.

**Comment by matthewmueller** [comment]
Hey @danshapir and @shtse8, thanks for the feedback! We've closed this issue because we've implemented the original use case, but the use case around references across tables could be a solid improvement that we'll consider separately. 

We've taken a note of your feedback in [this feedback issue](https://github.com/prisma/prisma/issues/15068). If you have any more suggestions, we'll be reviewing feedback in that issue. Thanks!
